mne/beamformer/_dics.py

```diff
@@ -1,32 +1,482 @@
+            'subject' : {string}
+                The subject ID.
+
+    See Also
+    --------
+    apply_dics, dics, dics_epochs
+
+    Notes
+    -----
+    The original reference is [1]_.
+
+    For more information about ``real_filter``, see the
+    `supplemental information <http://www.cell.com/cms/attachment/616681/4982593/mmc1.pdf>`_
+    from [3]_.
+
+    References
+    ----------
+    .. [1] Gross et al. Dynamic imaging of coherent sources: Studying neural
+           interactions in the human brain. PNAS (2001) vol. 98 (2) pp. 694-699
+    .. [2] Sekihara & Nagarajan. Adaptive spatial filters for electromagnetic
+           brain imaging (2008) Springer Science & Business Media
+    .. [3] Hipp JF, Engel AK, Siegel M (2011) Oscillatory Synchronization
+           in Large-Scale Cortical Networks Predicts Perception.
+           Neuron 69:387-396.
+    """  # noqa: E501
+    allowed_ori = [None, 'normal', 'max-power']
+    if pick_ori not in allowed_ori:
+        raise ValueError('"pick_ori" should be one of %s.' % allowed_ori)
+
+    frequencies = [np.mean(freq_bin) for freq_bin in csd.frequencies]
+    n_freqs = len(frequencies)
+    n_orient = forward['sol']['ncol'] // forward['nsource']
+
+    picks = _setup_picks(info=info, forward=forward)
+    _, ch_names, proj, vertices, G = _prepare_beamformer_input(
+        info, forward, label, picks=picks, pick_ori=pick_ori,
+        leadfield_norm=leadfield_norm,
+    )
+    csd_picks = [csd.names.index(ch) for ch in ch_names]
+
+    n_sources = G.shape[1] // n_orient
+
+    logger.info('Computing DICS spatial filters...')
+    Ws = []
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    computing DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_matrix(index=i)
+
+        if real_filter:
+            Cm = Cm.real
+
+        # Ensure the CSD is in the same order as the leadfield
+        Cm = Cm[csd_picks, :][:, csd_picks]
+
+        # Tikhonov regularization using reg parameter to control for
+        # trade-off between spatial resolution and noise sensitivity
+        # eq. 25 in Gross and Ioannides, 1999 Phys. Med. Biol. 44 2081
+        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+
+        # Compute spatial filters
+        W = np.dot(G.T, Cm_inv)
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            Gk = G[:, n_orient * k: n_orient * k + n_orient]
+
+            # Compute power at the source
+            Ck = np.dot(Wk, Gk)
+
+            # Normalize the spatial filters
+            if Wk.ndim == 2 and len(Wk) > 1:
+                # Free source orientation
+                # Officially, the formula calls for inv(Ck) @ Wk.
+                # However, since Ck is most likely rank deficient, the
+                # following is more stable in practice.
+                Wk /= np.diag(Ck)[:, np.newaxis]
```

Comment:
I definitely see the possible issues with the instability here, but I am not sure I understand taking the diagonal only - so I am curious :) 
I've quickly played around with it and it gives rather different results from taking the inverse (if Ck is not rank-deficient anyway) ... not sure if the off-diagonal elements are negligible?

Reply:
I will have to call in Jan Kujala to tell you more about this.

As far as my understanding goes, this choice was made based on extensive simulations and I guess, preference. The problem is, as you have encountered in the LCMV beamformer as well, that sometimes, the inverse of `Ck` does not exist.

How you have a choice of what to use instead of the inverse. We could add some regularization, drop 1 (or even 2) of the 3 ranks, or ignore the diagonal elements. These methods will produce different results, but since no true inverse exists to compare the results to, we'll have to evaluate which approach is better by evaluating the performance and stability of the resulting spatial filter as a whole.

But please don't take my word for this. My understanding of this implementation detail is limited.

Reply:
@wmvanvliet Please do follow up on this. Does Jan have a github account? ;-)

Reply:
He promised to reply on Monday.